pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,400 @@
+	oldNS := oldObj.(*v1.Namespace)
+	updatedNS := obj.(*v1.Namespace)
+	oldAnnotation, _ := oldNS.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	updatedAnnotation, _ := updatedNS.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	if oldAnnotation == updatedAnnotation {
+		return
+	}
+
+	klog.V(2).InfoS("Processing Namespace UPDATE event", "Namespace", klog.KObj(updatedNS))
+
+	affectedPods := l7c.getNonAnnotatedPodsFromNS(updatedNS)
+	for _, pod := range affectedPods {
+		podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+		l7c.queue.Add(podNN)
+	}
+}
+
+func (l7c *L7VisibilityController) getNonAnnotatedPodsFromNS(ns *v1.Namespace) []*v1.Pod {
+	var nonAnnotatedPods []*v1.Pod
+	pods, _ := l7c.podLister.Pods(ns.Name).List(labels.Everything())
+
+	// Only select the non annotated Pods, as annotated Pods are handled separately
+	for _, pod := range pods {
+		_, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+		if !ok {
+			nonAnnotatedPods = append(nonAnnotatedPods, pod)
+		}
+	}
+	return nonAnnotatedPods
+}
+
+func (l7c *L7VisibilityController) syncPod(podNN string) error {
+	podNamespace, podName := k8s.SplitNamespacedName(podNN)
+	pod, err := l7c.podLister.Pods(podNamespace).Get(podName)
+	if err != nil {
+		// Remove the TC flows if the Pod has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+	if !isValidPod(pod) {
+		return nil
+	}
+	annotationValue, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		podNS, err := l7c.namespaceLister.Get(pod.Namespace)
+		if err != nil {
+			// Remove TC flows if Namespace has been deleted
+			return l7c.removeTCFlow(podNN)
+		}
+		// Both Pod and Namespace are not annotated, remove the TC Mark flow
+		annotationValue, ok = podNS.Annotations[types.L7VisibilityAnnotationKey]
+		if !ok {
+			return l7c.removeTCFlow(podNN)
+		}
+	}
+
+	// Check if the annotation value is one of the specified values
+	var direction v1alpha2.Direction
+	annotationValue = strings.ToLower(annotationValue)
+	switch annotationValue {
+	case "ingress":
+		direction = v1alpha2.DirectionIngress
+	case "egress":
+		direction = v1alpha2.DirectionEgress
+	case "both":
+		direction = v1alpha2.DirectionBoth
+	default:
+		return errInvalidAnnotation
+	}
+	podInterfaces := l7c.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+	if len(podInterfaces) == 0 {
+		return errPodInterfaceNotFound
+	}
+	sourceOfPort := []uint32{uint32(podInterfaces[0].OFPort)}
+
+	// Start Suricata before starting traffic control mark flows
+	l7c.l7Reconciler.StartSuricataOnce()
+
+	oldDirection, exists := l7c.getMirroredDirection(podNN)
+	if exists {
+		if oldDirection == direction {
+			return nil
+		}
+		if err := l7c.removeTCFlow(podNN); err != nil {
+			return err
+		}
+	}
+	tcName := fmt.Sprintf("tcl7:%s", podNN)
+	if err := l7c.ofClient.InstallTrafficControlMarkFlows(tcName, sourceOfPort, l7c.targetPort, direction, v1alpha2.ActionMirror); err != nil {
```

Comment:
I have a doubt so I want to double check.
Looking at the OF implementation, it seems that we can only take a single traffic control flow for a given Pod. So if for a given Pod I have L7Visibility (with Mirroring action), L7NP (with Redirect action), and maybe a user-defined TC CR for IDS/IPS integration (with Redirect action), what will happen?
If I am right, this can be a bit of an issue: let's say I have L7NP enforced on a Pod and I enable L7 visibility. Does that mean that my L7NP will no longer be enforced, as traffic will now be mirrored instead of redirected?
I hope I am wrong but I need confirmation... @tnqn

Reply:
@antoninbas You are right unfortually. Currently all features relying these flows are mutually exclusive. I didn't have a good solution for it, as it wasn't clear what's the expected behavior when multiple actions are applied to the same Pod and it looked complicated to implement as well. Perhaps we could first document the limitation and rethink what's the right behavior, we can perhaps support applying at most one mirror action and one redirect action to a Pod and mirror happens before redirect.

Reply:
I am a bit concerned about the security implications of the current situation. The fact that an application developer can annotate a Pod to enable L7 visibility, and this can totally bypass L7NP enforcement and IDS/IPS is not a situation we want to be in IMO, even if L7Visibility is only Alpha. Is there a simple way to have lower priority for L7Visibility flows, so that everything else takes precedence? It seems that this is something we could do pretty easily. Additionally, if traffic is redirected to Suricata for L7NP or IDS/IPS, we would still get the HTTP events on the Suricata socket, so if L7Visibility is enabled on that Pod, we should still be able to export L7Visibility data, even if the L7Visibility flows do not actually take effect.

Maybe we could add the notion of priority to the TC flows API (system / high / medium / low), even if we don't expose it as part of the TrafficControl CRD (yet?).

Reply:
I have reviewed the flows for TrafficControl and L7NP. 

In `TrafficControlTable`, without the feature in this PR:

- The flow with `priorityHigh` is to match the returned packets of L7NP or TrafficControl redirect action.
- The flow with `priorityNormal` is to mark the packets selected by TrafficControl with `TrafficControlRedirectRegMark` or `TrafficControlMirrorRegMark`. 

```go
// Both for TrafficControl and L7NP, to match the returned packets from engine like Suricata, or something else.
TrafficControlTable.ofTable.BuildFlow(priorityHigh).
	Cookie(cookieID).
	MatchRegMark(OutputToOFPortRegMark, FromTCReturnRegMark).
	Action().GotoStage(stageOutput).
	Done(),


// For TrafficControl egress, to match the packets towards specific ofPorts.
TrafficControlTable.ofTable.BuildFlow(priorityNormal).
	Cookie(cookieID).
	MatchRegFieldWithValue(TargetOFPortField, port).
	Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
	Action().LoadRegMark(actionRegMark).
	Action().NextTable().
	Done()

// For TrafficControl ingress, to match the packets source from specific ofPorts.
TrafficControlTable.ofTable.BuildFlow(priorityNormal).
        Cookie(cookieID).
        MatchInPort(port).
        Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
        Action().LoadRegMark(actionRegMark).
        Action().NextTable().
        Done()
```

In `OutputTable`:
- The first two flows for TrafficControl are to match the packets marked in TrafficControlTable.
- The last flow for L7NP is to match the packets marked in https://github.com/antrea-io/antrea/blob/74e02ae73023247758d1e1cddf7a5a5598e18f51/pkg/agent/openflow/pipeline.go#L1857.

```go
// For TrafficControl.
OutputTable.ofTable.BuildFlow(priorityHigh+1).
	Cookie(cookieID).
	MatchRegMark(OutputToOFPortRegMark, TrafficControlMirrorRegMark).
	Action().OutputToRegField(TargetOFPortField).
	Action().OutputToRegField(TrafficControlTargetOFPortField).
	Done(),
OutputTable.ofTable.BuildFlow(priorityHigh+1).
	Cookie(cookieID).
	MatchRegMark(OutputToOFPortRegMark, TrafficControlRedirectRegMark).
	Action().OutputToRegField(TrafficControlTargetOFPortField).
	Done(),


// For L7NP.
OutputTable.ofTable.BuildFlow(priorityHigh+1).
	Cookie(cookieID).
	MatchRegMark(OutputToOFPortRegMark).
	MatchCTMark(L7NPRedirectCTMark).
	Action().PushVLAN(EtherTypeDot1q).
	Action().MoveRange(binding.NxmFieldCtLabel, binding.OxmFieldVLANVID, *L7NPRuleVlanIDCTLabel.GetRange(), *binding.VLANVIDRange).
	Action().Output(f.l7NetworkPolicyConfig.TargetOFPort).
	Done(),
```

From the existing flows:

L7NP and TrafficControl cannot be applied to the same Pod.  For example, a packet to a Pod, it has both `TrafficControlRedirectRegMark` and `L7NPRedirectCTMark`, in `OutputTable`, the packets can be matched by two different flows at the same chance. One flow is to redirect the packet to the ofPort set in TrafficControl CR, and another flow is to redirect the  packet to the ofPort used by L7NP Suricata engine. In short:
-  Without this PR, there is a chance to bypass L7NP enforcement by creating a TrafficControl applied to the same Pod.
- If users implement their own IDS/IPS design by using TrafficControl, creating another TrafficControl applied to the same Pod will not take effect since in TrafficControl controller code, it can be set as effect only when it is the only one TrafficControl CR applied to the Pod.

For this feature:
- `InstallTrafficControlMarkFlows` only installs the flows to mark the packets. The final mirroring action is performed in `OutputTable`. 
- It has the same issue mentioned above: there is a chance to bypass L7NP enforcement.

I will make a PR to fix the issue: by modifying the priority of the flow to match L7NP in `OutputTable` from `priorityHigh+1` to `priorityHigh+2`. This can ensure the L7NP enforcement. Note that, the TrafficControl CR applied to the same Pod will not work as expected util the L7NP applied to the Pod is deleted.

Another thing is that this feature could have a better way to implement its datapath. Currently, I supposed that @tushartathgur wanted to mirror the packets, then used the existing `InstallTrafficControlMarkFlows`. However, this is for TrafficControl and `featurePodConnectivity` according to the design of FlexiblePipeline.  Maybe we could implement it like this:


```go
func (c *client) InstallTrafficControlMarkFlows(name string, sourceOFPorts []uint32, targetOFPort uint32, direction crdv1alpha2.Direction, action crdv1alpha2.TrafficControlAction, feature string) error {
	if feature == "TrafficControl" {
		flows := c.featurePodConnectivity.trafficControlMarkFlows(sourceOFPorts, targetOFPort, direction, action)
		cacheKey := fmt.Sprintf("tc_%s", name)
		c.replayMutex.RLock()
		defer c.replayMutex.RUnlock()
		return c.modifyFlows(c.featurePodConnectivity.tcCachedFlows, cacheKey, flows)
	} else if feature == "L7Visibility" {
		flows := c.featureNetworkPolicy.l7VisibilityMarkFlows(sourceOFPorts, targetOFPort, direction, action)
		cacheKey := fmt.Sprintf("tc_%s", name)
		c.replayMutex.RLock()
		defer c.replayMutex.RUnlock()
		return c.modifyFlows(c.featurePodConnectivity.tcCachedFlows, cacheKey, flows)
	}
	return nil
}

// trafficControlMarkFlows generates the flows to mark the packets that need to be redirected or mirrored.
func (f *featurePodConnectivity) l7VisibilityMarkFlows(sourceOFPorts []uint32, targetOFPort uint32) []binding.Flow {
	cookieID := f.cookieAllocator.Request(f.category).Raw()
	var flows []binding.Flow
	for _, port := range sourceOFPorts {
		flows = append(flows, TrafficControlTable.ofTable.BuildFlow(priorityLow).
			Cookie(cookieID).
			MatchRegFieldWithValue(TargetOFPortField, port).
			Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
			Action().LoadRegMark(TrafficControlMirrorRegMark).
			Action().NextTable().
			Done())
		flows = append(flows, TrafficControlTable.ofTable.BuildFlow(priorityLow).
			Cookie(cookieID).
			MatchInPort(port).
			Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
			Action().LoadRegMark(TrafficControlMirrorRegMark).
			Action().NextTable().
			Done())
	}
	return flows
}

```

Or we could just add a new method to ofClient like `InstallL7VisibilityMarkFlows`